internal/olm/operator/registry/index_image.go

```diff
@@ -124,14 +151,319 @@ func (c IndexImageCatalogCreator) CreateCatalog(ctx context.Context, name string
+		return "", err
+	}
+
+	if content == "" {
+		return "", errors.New("File-Based Catalog contents cannot be empty")
+	}
+
+	if err = ValidateFBC(originalDeclCfg); err != nil {
+		log.Errorf("error validating the generated FBC: %v", err)
+		return "", err
+	}
+
+	log.Infof("Generated a valid Upgraded File-Based Catalog")
+
+	return content, nil
+}
+
+// setupFBCupdates starts the process of upgrading a bundle in an FBC. This function will recreate the FBC that was generated
+// during run bundle and upgrade a specific bundle in the specified channel.
+func setupFBCupdates(c *IndexImageCatalogCreator, ctx context.Context) error {
+	var (
+		originalDeclcfg *declarativeconfig.DeclarativeConfig
+		err             error
+		render          action.Render
+	)
+
+	// render the index image if it is not the default index image
+	render = action.Render{Refs: []string{}}
+	if c.IndexImage != DefaultIndexImage {
+		render.Refs = append(render.Refs, c.IndexImage)
+	}
+
+	log.SetOutput(ioutil.Discard)
+	originalDeclcfg, err = render.Run(ctx)
+	log.SetOutput(os.Stdout)
+
+	if err != nil {
+		return err
+	}
+
+	// add the upgraded bundle to the list of previous bundles so that they can be rendered with a single API call
+	c.PreviousBundles = append(c.PreviousBundles, c.BundleImage)
+	f := &FBCContext{
+		Package:     c.PackageName,
+		Refs:        c.PreviousBundles,
+		ChannelName: c.ChannelName,
+	}
+
+	// Adding the FBC "f" to the originalDeclcfg to generate a new FBC
+	declcfg, err := upgradeFBC(f, originalDeclcfg, ctx)
+	if err != nil {
+		log.Errorf("error creating the upgraded FBC: %v", err)
+		return err
+	}
+
+	// convert declarative config to string
+	content, err := StringifyDeclConfig(declcfg)
+
+	if err != nil {
+		log.Errorf("error converting declarative config to string: %v", err)
+		return err
+	}
+
+	if content == "" {
+		return errors.New("File-Based Catalog contents cannot be empty")
+	}
+
+	// validate the declarative config
+	if err = ValidateFBC(declcfg); err != nil {
+		log.Errorf("error validating the generated FBC: %v", err)
+		return err
+	}
+
+	log.Infof("Generated a valid Upgraded File-Based Catalog")
+
+	c.FBCContent = content
+
+	return nil
+}
+
+// upgradeFBC constructs a new File-Based Catalog from both the FBCConext object and the declarative config object. This function will check to see
+// if the FBCContext object "f" is already present in the original declarative config.
+func upgradeFBC(f *FBCContext, originalDeclCfg *declarativeconfig.DeclarativeConfig, ctx context.Context) (*declarativeconfig.DeclarativeConfig, error) {
+	var (
+		declcfg *declarativeconfig.DeclarativeConfig
+		err     error
+	)
+
+	// Rendering the bundle image and index image into declarative config format
+	render := action.Render{
+		Refs: f.Refs,
+	}
+
+	log.SetOutput(ioutil.Discard)
+	declcfg, err = render.Run(ctx)
+	log.SetOutput(os.Stdout)
+
+	if err != nil {
+		log.Errorf("error in rendering the bundle and index image: %v", err)
```

Comment:
It looks like we are logging these errors twice. Can we instead return an error with additional information in the log message by doing `return fmt.Errorf`?

The same can be applied to other lines of code where `log.Error` is being called.